internal/objectstore/remote/retriever.go

```diff
@@ -0,0 +1,298 @@
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		r.stopAllRetrievers(ctx, retrievers)
+	}()
+
+	// We want to run it like this so we can return the first successful result
+	// and close the other readers. If we use for range over the channel, we
+	// have no way to close the result.
+	results := make(map[int]struct{})
+	for {
+		select {
+		case <-r.tomb.Dying():
+			return nil, -1, tomb.ErrDying
+
+		case <-ctx.Done():
+			return nil, -1, ctx.Err()
+
+		case res := <-result:
+			results[res.index] = struct{}{}
+
+			// If the blob is not found on that remote, continue to the next one
+			// until it is exhausted. This is a race to find it first.
+			if err := res.err; errors.Is(err, BlobNotFound) {
+				if len(results) == len(remotes) {
+					return nil, -1, BlobNotFound
+				}
+				continue
+			} else if err != nil {
+				// If there is an error that is not BlobNotFound, return it
+				return nil, -1, err
+			}
+
+			// Stop all the other retrievers, we don't want to cancel the
+			// retriever that is currently being used, as that will cause the
+			// reader to be closed.
+			for _, retriever := range retrievers {
+				if retriever.index == res.index {
+					continue
+				}
+
+				retriever.Kill()
+
+				// Don't wait for them to stop, when closing the reader, we will
+				// wait for them to stop then. We just want them to stop
+				// processing the request as soon as possible.
+			}
+
+			return &retrieverReaderCloser{
+				reader: res.reader,
+				closer: func() {
+					r.stopAllRetrievers(ctx, retrievers)
+				},
+			}, res.size, nil
+		}
+	}
+}
+
+// Kill stops the BlobRetriever.
+func (r *BlobRetriever) Kill() {
+	r.tomb.Kill(nil)
+}
+
+// Wait waits for the BlobRetriever to stop.
+func (r *BlobRetriever) Wait() error {
+	return r.tomb.Wait()
+}
+
+func (r *BlobRetriever) loop() error {
+	select {
+	case <-r.tomb.Dying():
+		return tomb.ErrDying
+	}
+}
+
+// stopAllRetrievers stops all the retrievers and waits for them to stop. This
+// ensures that there are no dangling goroutines.
+func (r *BlobRetriever) stopAllRetrievers(ctx context.Context, retrievers []*retriever) {
+	// Kill 'Em All.
+	for _, retriever := range retrievers {
+		retriever.Kill()
+		// Wait for them to stop, we don't want to leave any hanging.
+		if err := retriever.Wait(); err != nil {
+			r.logger.Errorf(ctx, "failed to stop blob retriever: %v", err)
+		}
+	}
+}
+
+type retriever struct {
+	tomb tomb.Tomb
+
+	index           int
+	remote          apiremotecaller.RemoteConnection
+	newObjectClient NewObjectClientFunc
+	clock           clock.Clock
+	logger          logger.Logger
+}
+
+func newRetriever(index int, remote apiremotecaller.RemoteConnection, newObjectClient NewObjectClientFunc, clock clock.Clock, logger logger.Logger) *retriever {
+	t := &retriever{
+		index:           index,
+		remote:          remote,
+		newObjectClient: newObjectClient,
+		clock:           clock,
+		logger:          logger,
+	}
+
+	t.tomb.Go(func() error {
```

Comment:
Why does this have a tomb?

Reply:
Each objectstore will have a retriever, and upon closing the the objectstore, we also need to ensure that everything is stopped and more importantly closed and finished, which is the semantic difference between context and tomb.

Reply:
Yeah I got that, just it wasn't being used :)